examples/cms/src/FieldMapping.tsx

```diff
@@ -0,0 +1,233 @@
+                        event.stopPropagation()
+                    }}
+                />
+                <span>{field.id}</span>
+            </div>
+            <ChevronIcon />
+            <input
+                type="text"
+                style={{
+                    width: "100%",
+                    opacity: isIgnored ? 0.5 : 1,
+                }}
+                disabled={isIgnored}
+                placeholder={field.id}
+                value={hasCustomName ? field.name : ""}
+                onChange={event => {
+                    const value = event.target.value
+                    if (!value.trim()) {
+                        setHasCustomName(false)
+                        onFieldNameChange(field.id, field.id)
+                    } else {
+                        setHasCustomName(true)
+                        onFieldNameChange(field.id, value.trimStart())
+                    }
+                }}
+            />
+        </>
+    )
+})
+
+interface FieldMappingProps {
+    collection: ManagedCollection
+    dataSource: DataSource
+    initialSlugFieldId: string | null
+}
+
+type FieldMappingStatus = "mapping-fields" | "loading-fields" | "syncing-collection"
+
+export function FieldMapping({ collection, dataSource, initialSlugFieldId }: FieldMappingProps) {
+    const [status, setStatus] = useState<FieldMappingStatus>("loading-fields")
+    const isSyncing = status === "syncing-collection"
+    const isLoadingFields = status === "loading-fields"
+
+    const sourceFields = useMemo(() => computeFieldsFromDataSource(dataSource), [dataSource])
+    const possibleSlugFields = useMemo(() => sourceFields.filter(field => field.type === "string"), [sourceFields])
+
+    const [selectedSlugField, setSelectedSlugField] = useState<ManagedCollectionField | null>(
+        possibleSlugFields.find(field => field.id === initialSlugFieldId) ?? possibleSlugFields[0] ?? null
+    )
+
+    const [fields, setFields] = useState<ManagedCollectionField[]>([])
+    const [ignoredFieldIds, setIgnoredFieldIds] = useState<Set<string>>(new Set())
+
+    useEffect(() => {
+        collection
+            .getFields()
+            .then(fields => {
+                setFields(mergeFieldsWithExistingFields(sourceFields, fields))
+
+                if (Boolean(initialSlugFieldId) === false && fields.length === 0) {
+                    return
+                }
+
+                const ignoredFields = sourceFields.filter(
+                    field => !fields.some(existingField => existingField.id === field.id)
+                )
+
+                setIgnoredFieldIds(new Set(ignoredFields.map(field => field.id)))
+            })
+            .catch(error => {
+                console.error(error)
+                framer.notify("Failed to load existing fields.", { variant: "error" })
+            })
+            .finally(() => {
+                setStatus("mapping-fields")
+            })
+    }, [collection, sourceFields, initialSlugFieldId])
```

Comment:
This effect can run multiple times and it is doing something async, so we need to be able to cancel it when the effect is cleaned up